moshi/src/main/java/com/squareup/moshi/Types.java

```diff
@@ -284,11 +284,21 @@ public static Type collectionElementType(Type context, Class<?> contextRawType)
     Type mapType = getSupertype(context, contextRawType, Map.class);
     if (mapType instanceof ParameterizedType) {
       ParameterizedType mapParameterizedType = (ParameterizedType) mapType;
-      return mapParameterizedType.getActualTypeArguments();
+      return getCanonicalizedMapTypeArguments(mapParameterizedType);
```

Comment:
Frustratingly, this isn’t even enough. We need to do likewise above for collections, as on 270.